README.md
# snappystream

A Node.js library for supporting the
[Snappy](https://code.google.com/p/snappy/) framing format via streams. See
the [Snappy Framing Format
Description](https://github.com/google/snappy/blob/master/framing_format.txt) for
details.

Usage:
SnappyStream and UnsnappyStream are
[Transform streams](http://nodejs.org/api/stream.html#stream_class_stream_transform).

```javascript
const {SnappyStream} = require('snappystream')
const fs = require('fs')

const inStream = fs.createReadStream('snappy.txt')
const snappyStream = new SnappyStream()
const outStream = fs.createWriteStream('snappy_frame.txt')

inStream.pipe(snappyStream).pipe(outStream)
```

UnsnappyStream constructor takes an optional argument `verifyChecksums`
which is false by default.

```javascript
const {UnsnappyStream} = require('snappysteam')
const fs = require('fs')

const inStream = fs.createReadStream('snappy_frame.txt')
const unsnappyStream = new UnsnappyStream(true)

unsnappyStream.on('end', function () {
  console.log(unsnappyStream.read())
})

inStream.pipe(unsnappyStream)
```


CHANGELOG.md
# snappystream

## 2.1.1

### Patch Changes

- e859a41: Actually remove int24

## 2.1.0

### Minor Changes

- 4f72e7f: Dropped int24 dependency
- 4f72e7f: Migrate to typescript


test/snappystream.test.js
const snappy = require('snappy')
const {SnappyStream} = require('../lib/snappystreams')

const sentence = 'the quick brown fox jumped over the lazy dog.'
const txt = [sentence, sentence, sentence].join('\n')

// Generate a snappy stream from data. Return the snappy stream as a string.
async function compress(data) {
  let frames = []
  const compressor = new SnappyStream()
  compressor.write(data)
  compressor.end()

  for await (const frame of compressor) {
    frames.push(frame)
  }
  return Buffer.concat(frames)
}

describe('SnappyStream', () => {
  describe('stream identifier', () => {
    let compressedFrames
    beforeAll(async () => {
      compressedFrames = await compress(txt)
    })

    it('should have the stream identifier chunk ID', () => {
      expect(compressedFrames.readUInt8(0)).toBe(0xff)
    })

    it('should have the stream identifier chunk size of 6 bytes', () => {
      expect(compressedFrames.readUIntLE(1, 3)).toBe(6)
    })

    it('should have the stream identifier payload', () => {
      expect(compressedFrames.slice(4, 10).toString()).toBe('sNaPpY')
    })
  })

  describe('single compressed frame', () => {
    let compressedData
    let compressedFrames
    beforeAll(async () => {
      compressedData = await snappy.compress(txt)
      compressedFrames = (await compress(txt)).subarray(10)
    })

    it('should start with the compressed data chunk ID', () => {
      expect(compressedFrames.readUInt8(0)).toBe(0x00)
    })

    it('should have a valid frame size', () => {
      // Frame size is the size of the checksum mask (4 bytes) and the byte
      // length of the snappy compressed data.
      const frameSize = 4 + compressedData.length
      expect(compressedFrames.readUIntLE(1, 3)).toBe(frameSize)
    })

    it('should have a valid checksum mask', () => {
      expect(compressedFrames.readUInt32LE(4)).toBe(0x63885e6b)
    })

    it('should have match decompressed data', async () => {
      const payload = compressedFrames.subarray(8)
      const uncompressedPayload = await snappy.uncompress(payload)
      expect(uncompressedPayload.toString()).toBe(txt)
    })
  })

  describe('multiple compressed frames', () => {
    let data
    let compressedFrames
    beforeAll(async () => {
      // Two frames worth of data.
      data = new Array(100000).join('a')
      compressedFrames = (await compress(data)).subarray(10)
    })

    it('should have the first chunk start with a compressed data chunk ID', () => {
      expect(compressedFrames.readUInt8(0)).toBe(0x00)
    })

    it('should have the 1st chunk with an uncompressed size of 65,536', async () => {
      const frameSize = compressedFrames.readUIntLE(1, 3)
      const compressedData = compressedFrames.subarray(8, frameSize + 4)

      const frameData = await snappy.uncompress(compressedData)
      expect(frameData.length).toBe(65536)
      expect(frameData.toString()).toBe(data.slice(0, 65536))
    })

    it('should have the 2nd chunk start with a compressed data chunk ID', () => {
      expect(compressedFrames.readUInt8(3085)).toBe(0x00)
    })

    it('should have the 2nd chunk with an uncompressed size of 34,464', async () => {
      const secondFrame = compressedFrames.subarray(3085)
      const frameSize = secondFrame.readUIntLE(1, 3)
      expect(frameSize).toBe(secondFrame.length - 4)

      const frameData = await snappy.uncompress(secondFrame.slice(8))
      expect(frameData.toString()).toEqual(data.slice(65536))
    })
  })
})


test/snappystream.test.ts
import snappy from 'snappy'
import {SnappyStream} from '../lib/snappystreams'

const sentence = 'the quick brown fox jumped over the lazy dog.'
const txt = [sentence, sentence, sentence].join('\n')

// Generate a snappy stream from data. Return the snappy stream as a string.
async function compress(data: string) {
  const frames: Buffer[] = []
  const compressor = new SnappyStream()
  compressor.write(data)
  compressor.end()

  for await (const frame of compressor) {
    frames.push(frame as Buffer)
  }
  return Buffer.concat(frames)
}

describe('SnappyStream', () => {
  describe('stream identifier', () => {
    let compressedFrames: Buffer
    beforeAll(async () => {
      compressedFrames = await compress(txt)
    })

    it('should have the stream identifier chunk ID', () => {
      expect(compressedFrames.readUInt8(0)).toBe(0xff)
    })

    it('should have the stream identifier chunk size of 6 bytes', () => {
      expect(compressedFrames.readUIntLE(1, 3)).toBe(6)
    })

    it('should have the stream identifier payload', () => {
      expect(compressedFrames.slice(4, 10).toString()).toBe('sNaPpY')
    })
  })

  describe('single compressed frame', () => {
    let compressedData: Buffer
    let compressedFrames: Buffer
    beforeAll(async () => {
      compressedData = await snappy.compress(txt)
      compressedFrames = (await compress(txt)).subarray(10)
    })

    it('should start with the compressed data chunk ID', () => {
      expect(compressedFrames.readUInt8(0)).toBe(0x00)
    })

    it('should have a valid frame size', () => {
      // Frame size is the size of the checksum mask (4 bytes) and the byte
      // length of the snappy compressed data.
      const frameSize = 4 + compressedData.length
      expect(compressedFrames.readUIntLE(1, 3)).toBe(frameSize)
    })

    it('should have a valid checksum mask', () => {
      expect(compressedFrames.readUInt32LE(4)).toBe(0x63885e6b)
    })

    it('should have match decompressed data', async () => {
      const payload = compressedFrames.subarray(8)
      const uncompressedPayload = await snappy.uncompress(payload)
      expect(uncompressedPayload.toString()).toBe(txt)
    })
  })

  describe('multiple compressed frames', () => {
    let data: string
    let compressedFrames: Buffer
    beforeAll(async () => {
      // Two frames worth of data.
      data = new Array(100000).join('a')
      compressedFrames = (await compress(data)).subarray(10)
    })

    it('should have the first chunk start with a compressed data chunk ID', () => {
      expect(compressedFrames.readUInt8(0)).toBe(0x00)
    })

    it('should have the 1st chunk with an uncompressed size of 65,536', async () => {
      const frameSize = compressedFrames.readUIntLE(1, 3)
      const compressedData = compressedFrames.subarray(8, frameSize + 4)

      const frameData = await snappy.uncompress(compressedData)
      expect(frameData.length).toBe(65536)
      expect(frameData.toString()).toBe(data.slice(0, 65536))
    })

    it('should have the 2nd chunk start with a compressed data chunk ID', () => {
      expect(compressedFrames.readUInt8(3085)).toBe(0x00)
    })

    it('should have the 2nd chunk with an uncompressed size of 34,464', async () => {
      const secondFrame = compressedFrames.subarray(3085)
      const frameSize = secondFrame.readUIntLE(1, 3)
      expect(frameSize).toBe(secondFrame.length - 4)

      const frameData = await snappy.uncompress(secondFrame.subarray(8))
      expect(frameData.toString()).toEqual(data.slice(65536))
    })
  })
})


test/snappystream_test.js
require('should')
const int24 = require('int24')
const snappy = require('snappy')
const { SnappyStream } = require('../lib/snappystreams')

const sentence = 'the quick brown fox jumped over the lazy dog.'
const txt = [sentence, sentence, sentence].join('\n')

// Generate a snappy stream from data. Return the snappy stream as a string.
function compress (data, callback) {
  let compressedFrames = Buffer.alloc(0)
  const compressor = new SnappyStream()

  compressor.on('readable', () => {
    data = compressor.read()
    if (!data) {
      return
    }

    return (compressedFrames = Buffer.concat([compressedFrames, data]))
  })
  compressor.on('end', () => callback(null, compressedFrames))

  compressor.write(data)
  return compressor.end()
}

describe('SnappyStream', () => {
  describe('stream identifer', () => {
    let compressedFrames = null

    before(done => {
      compress(txt, (err, data) => {
        compressedFrames = data
        return done(err)
      })
    })

    it('should have the stream identifier chunk ID', () =>
      compressedFrames.readUInt8(0).should.eql(0xff))

    it('should have the stream identifer chunk size of 6 bytes', () =>
      int24.readUInt24LE(compressedFrames, 1).should.eql(6))

    return it('should have the stream identifier payload', () =>
      compressedFrames
        .slice(4, 10)
        .toString()
        .should.eql('sNaPpY'))
  })

  describe('single compressed frame', () => {
    let compressedFrames = null
    let compressedData = null

    before(done =>
      snappy.compress(txt, (err, snappyData) => {
        if (err) { return done(err) }
        compressedData = snappyData

        return compress(txt, (err, out) => {
          compressedFrames = out.slice(10)
          return done(err)
        })
      })
    )

    it('should start with the compressed data chunk ID', () =>
      compressedFrames.readUInt8(0).should.eql(0x00))

    it('should have a valid frame size', () => {
      // Frame size is the size of the checksum mask (4 bytes) and the byte
      // length of the snappy compressed data.
      const frameSize = 4 + compressedData.length
      return int24.readUInt24LE(compressedFrames, 1).should.eql(frameSize)
    })

    it('should have a valid checksum mask', () =>
      compressedFrames.readUInt32LE(4).should.eql(0x63885e6b))

    return it('should have match decompressed data', done => {
      const payload = compressedFrames.slice(8)
      return snappy.uncompress(payload, { asBuffer: false },
        (err, uncompressedPayload) => {
          uncompressedPayload.should.eql(txt)
          return done(err)
        })
    })
  })

  return describe('multiple compressed frames', () => {
    // Two frames worth of data.
    const data = new Array(100000).join('a')
    let compressedFrames = Buffer.alloc(0)

    before(done =>
      compress(data, (err, compressedData) => {
        compressedFrames = compressedData.slice(10)
        return done(err)
      })
    )

    it('should have the first chunk start with a compressed data chunk ID', () =>
      compressedFrames.readUInt8(0).should.eql(0x00))

    it('should have the 1st chunk with an uncompressed size of 65,536', done => {
      const frameSize = int24.readUInt24LE(compressedFrames, 1)
      const compressedData = compressedFrames.slice(8, frameSize + 4)

      return snappy.uncompress(compressedData, (err, frameData) => {
        if (err) {
          return done(err)
        }

        frameData.length.should.eql(65536)
        frameData.toString().should.eql(data.slice(0, 65536))
        return done()
      })
    })

    it('should have the 2nd chunk start with a compressed data chunk ID', () =>
      compressedFrames.readUInt8(3085).should.eql(0x00))

    return it('should have the 2nd chunk with an uncompressed size of 34,464',
      done => {
        const secondFrame = compressedFrames.slice(3085)
        const frameSize = int24.readUInt24LE(secondFrame, 1)

        frameSize.should.eql(secondFrame.length - 4)

        return snappy.uncompress(
          secondFrame.slice(8),
          { asBuffer: false },
          function (err, frameData) {
            if (err) {
              return done(err)
            }

            frameData.should.eql(data.slice(65536))
            return done()
          }
        )
      })
  })
})


lib/snappystreams.js
const {compress, uncompress} = require('snappy')
const {crc32c} = require('@node-rs/crc32')
const stream = require('stream')

const CHUNKS = {
  streamIdentifier: 0xff,
  compressedData: 0x00,
  uncompressedData: 0x01,
  padding: 0xfe,
  unskippable: (v) => v >= 0x02 && v <= 0x7f,
}

const STREAM_IDENTIFIER = Buffer.from([
  0xff, 0x06, 0x00, 0x00, 0x73, 0x4e, 0x61, 0x50, 0x70, 0x59,
])
const MAX_FRAME_DATA_SIZE = 65536

const checksumMask = function (data) {
  const c = crc32c(data)
  const result = new Uint32Array(1)
  result[0] = ((c >>> 15) | (c * Math.pow(2, 17))) + 0xa282ead8
  return result[0]
}

class SnappyStream extends stream.Transform {
  constructor(options) {
    super(options)
    this.push(STREAM_IDENTIFIER)
  }

  // No buffering of data before producing a compressed frame. If the data size
  // exceeds the size of a frame, then it will automatically be split across
  // frames per the Snappy frame spec.
  _transform(data, encoding, callback) {
    // Split data if need be into chunks no larger than the maximum size for
    // a frame.
    const out = Buffer.from(data)

    const dataChunks = []
    for (let offset = 0; offset < out.length / MAX_FRAME_DATA_SIZE; offset++) {
      const start = offset * MAX_FRAME_DATA_SIZE
      const end = start + MAX_FRAME_DATA_SIZE
      dataChunks.push(out.subarray(start, end))
    }
    Promise.all(dataChunks.map(chunk => compress(chunk)))
      .then((compressedDataChunks) => {
        const frameChunks = []
        for (let i = 0; i < dataChunks.length; i++) {
          const chunkData = dataChunks[i]
          const frameData = compressedDataChunks[i]

          const frameStart = Buffer.alloc(8)

          let headerType = CHUNKS.compressedData
          let payload = frameData

          // If the improvement isn't more than 12.5% then use uncompressed
          // data.
          if (frameData.length >= chunkData.length - chunkData.length / 8) {
            headerType = CHUNKS.uncompressedData
            payload = chunkData
          }

          frameStart.writeUInt8(headerType, 0)
          frameStart.writeUintLE(payload.length + 4, 1, 3)
          frameStart.writeUInt32LE(checksumMask(chunkData), 4)

          frameChunks.push(frameStart)
          frameChunks.push(payload)
        }

        this.push(Buffer.concat(frameChunks))
        return callback()
      })
      .catch((err) => {
        callback(err)
      })
  }
}

class UnsnappyStream extends stream.Transform {
  constructor(verifyChecksums = false, options = {}) {
    super(options)
    this.verifyChecksums = verifyChecksums
    this.identifierFound = false
    this.frameBuffer = null
  }

  // Returns snappy compressed payload. Throws an error if the checksum fails
  // provided stream is checking checksums.
  framePayload(frame) {
    const frameLength = frame.readUIntLE(1, 3)
    return frame.subarray(8, frameLength + 4)
  }

  frameMask(frame) {
    return frame.readUInt32LE(4)
  }

  // Data contains at least one full frame.
  hasFrame(data) {
    return data.length > 4 && data.readUIntLE(1, 3) + 4 <= data.length
  }

  // Return the buffer starting at the next frame. It assumes that a full frame
  // exists within data.
  toNextFrame(data) {
    const frameLength = data.readUIntLE(1, 3)
    return data.subarray(4 + frameLength)
  }

  verify(mask, data) {
    if (this.verifyChecksums && checksumMask(data) !== mask) {
      return Promise.reject(new Error('Frame failed checksum'))
    }

    return Promise.resolve(data)
  }

  processChunks(chunks, done = () => {}) {
    const verify = this.verify.bind(this)
    return Promise.all(
      chunks.map(async (chunk) => {
        const [frameType, mask, payload] = chunk
        const uncompressedData = frameType === CHUNKS.uncompressedData ? payload : await uncompress(payload)
        await verify(mask, uncompressedData)
        return uncompressedData
      })
    ).then((data) => {
      this.push(Buffer.concat(data))
      done()
    }).catch((err) => {
      return this.emit('error', err)
    })
  }

  _transform(data, encoding, done) {
    // Tuples of frame ID and frame payload
    const chunks = []

    if (encoding) {
      data = Buffer.from(data, encoding)
    }
    if (this.frameBuffer) {
      data = Buffer.concat([this.frameBuffer, data])
    }
    this.frameBuffer = null

    if (
      !this.identifierFound &&
      data.readUInt8(0) !== CHUNKS.streamIdentifier
    ) {
      return this.emit('error', new Error('Missing snappy stream identifier'))
    }

    // Loop only while a full frame is available within data.
    while (this.hasFrame(data)) {
      const frameId = data.readUInt8(0)

      try {
        switch (frameId) {
          case CHUNKS.streamIdentifier:
            if (data.slice(0, 10).toString() !== STREAM_IDENTIFIER.toString()) {
              return this.emit('error', new Error('Invalid stream identifier'))
            }
            this.identifierFound = true
            break
          case CHUNKS.compressedData:
            chunks.push([
              CHUNKS.compressedData,
              this.frameMask(data),
              this.framePayload(data),
            ])
            break
          case CHUNKS.uncompressedData:
            chunks.push([
              CHUNKS.uncompressedData,
              this.frameMask(data),
              this.framePayload(data),
            ])
            break
          case CHUNKS.unskippable(frameId):
            return this.emit('error', new Error('Encountered unskippable frame'))
        }
      } catch (err) {
        return this.emit('error', err)
      }

      data = this.toNextFrame(data)
    }

    if (data.length) {
      this.frameBuffer = data
    }

    if (chunks.length) {
      return this.processChunks(chunks, done)
    } else {
      return done()
    }
  }

  _flush(done) {
    if (this.frameBuffer && this.frameBuffer.length) {
      done(new Error('Failed to decompress Snappy stream'))
    }
    done()
  }
}

module.exports = {SnappyStream, UnsnappyStream}


lib/snappystreams.ts
import {compress, uncompress} from 'snappy'
import {crc32c} from '@node-rs/crc32'
import stream from 'stream'

const CHUNKS = {
  streamIdentifier: 0xff,
  compressedData: 0x00,
  uncompressedData: 0x01,
  padding: 0xfe,
  unskippable: (v: number) => (v >= 0x02 && v <= 0x7f ? v : -1),
}

const STREAM_IDENTIFIER = Buffer.from([
  0xff, 0x06, 0x00, 0x00, 0x73, 0x4e, 0x61, 0x50, 0x70, 0x59,
])
const MAX_FRAME_DATA_SIZE = 65536

const checksumMask = (data: Buffer | string) => {
  const c = crc32c(data)
  const result = new Uint32Array(1)
  result[0] = ((c >>> 15) | (c * Math.pow(2, 17))) + 0xa282ead8
  return result[0]
}

export class SnappyStream extends stream.Transform {
  constructor(options?: stream.TransformOptions) {
    super(options)
    this.push(STREAM_IDENTIFIER)
  }

  // No buffering of data before producing a compressed frame. If the data size
  // exceeds the size of a frame, then it will automatically be split across
  // frames per the Snappy frame spec.
  _transform(
    data: Buffer | string,
    _: BufferEncoding,
    callback: (err?: Error) => void
  ) {
    // Split data if need be into chunks no larger than the maximum size for
    // a frame.
    const out = Buffer.from(data)

    const dataChunks: Buffer[] = []
    for (let offset = 0; offset < out.length / MAX_FRAME_DATA_SIZE; offset++) {
      const start = offset * MAX_FRAME_DATA_SIZE
      const end = start + MAX_FRAME_DATA_SIZE
      dataChunks.push(out.subarray(start, end))
    }
    Promise.all(dataChunks.map((chunk) => compress(chunk)))
      .then((compressedDataChunks) => {
        const frameChunks: Buffer[] = []
        for (let i = 0; i < dataChunks.length; i++) {
          const chunkData = dataChunks[i]!
          const frameData = compressedDataChunks[i]!

          const frameStart = Buffer.alloc(8)

          let headerType = CHUNKS.compressedData
          let payload = frameData

          // If the improvement isn't more than 12.5% then use uncompressed
          // data.
          if (frameData.length >= chunkData.length - chunkData.length / 8) {
            headerType = CHUNKS.uncompressedData
            payload = chunkData
          }

          frameStart.writeUInt8(headerType, 0)
          frameStart.writeUintLE(payload.length + 4, 1, 3)
          frameStart.writeUInt32LE(checksumMask(chunkData), 4)

          frameChunks.push(frameStart)
          frameChunks.push(payload)
        }

        this.push(Buffer.concat(frameChunks))
        return callback()
      })
      .catch((err: Error) => {
        callback(err)
      })
  }
}

export class UnsnappyStream extends stream.Transform {
  public verifyChecksums: boolean
  private identifierFound: boolean
  private frameBuffer: Buffer | null

  constructor(verifyChecksums = false, options = {}) {
    super(options)
    this.verifyChecksums = verifyChecksums
    this.identifierFound = false
    this.frameBuffer = null
  }

  // Returns snappy compressed payload. Throws an error if the checksum fails
  // provided stream is checking checksums.
  framePayload(frame: Buffer) {
    const frameLength = frame.readUIntLE(1, 3)
    return frame.subarray(8, frameLength + 4)
  }

  frameMask(frame: Buffer) {
    return frame.readUInt32LE(4)
  }

  // Data contains at least one full frame.
  hasFrame(data: Buffer) {
    return data.length > 4 && data.readUIntLE(1, 3) + 4 <= data.length
  }

  // Return the buffer starting at the next frame. It assumes that a full frame
  // exists within data.
  toNextFrame(data: Buffer) {
    const frameLength = data.readUIntLE(1, 3)
    return data.subarray(4 + frameLength)
  }

  verify(mask: number | null, data: Buffer | string) {
    if (this.verifyChecksums && checksumMask(data) !== mask) {
      throw new Error('Frame failed checksum')
    }
  }

  async processChunks(
    chunks: [number, number | null, Buffer][],
    done?: (err?: Error) => void
  ) {
    const verify = this.verify.bind(this)
    try {
      const uncompressFrames = await Promise.all(
        chunks.map(async (chunk) => {
          const [frameType, mask, payload] = chunk
          const uncompressedData =
            frameType === CHUNKS.uncompressedData
              ? payload
              : await uncompress(payload)
          verify(mask, uncompressedData)
          return Buffer.isBuffer(uncompressedData)
            ? uncompressedData
            : Buffer.from(uncompressedData)
        })
      )
      this.push(Buffer.concat(uncompressFrames))
      if (done) {
        done()
      }
    } catch (err) {
      this.emit('error', err)
    }
  }

  _transform(
    data: Buffer | string,
    encoding: BufferEncoding | null,
    done: (err?: Error) => void
  ) {
    // Tuples of frame ID and frame payload
    const chunks: [number, number | null, Buffer][] = []

    let _data = encoding
      ? Buffer.from(data as string, encoding)
      : (data as Buffer)
    if (this.frameBuffer) {
      _data = Buffer.concat([this.frameBuffer, _data])
    }
    this.frameBuffer = null

    if (
      !this.identifierFound &&
      _data.readUInt8(0) !== CHUNKS.streamIdentifier
    ) {
      return this.emit('error', new Error('Missing snappy stream identifier'))
    }

    // Loop only while a full frame is available within data.
    while (this.hasFrame(_data)) {
      const frameId = _data.readUInt8(0)

      try {
        switch (frameId) {
          case CHUNKS.streamIdentifier:
            if (
              _data.subarray(0, 10).toString() !== STREAM_IDENTIFIER.toString()
            ) {
              return this.emit('error', new Error('Invalid stream identifier'))
            }
            this.identifierFound = true
            break
          case CHUNKS.compressedData:
            chunks.push([
              CHUNKS.compressedData,
              this.frameMask(_data),
              this.framePayload(_data),
            ])
            break
          case CHUNKS.uncompressedData:
            chunks.push([
              CHUNKS.uncompressedData,
              this.frameMask(_data),
              this.framePayload(_data),
            ])
            break
          case CHUNKS.unskippable(frameId):
            return this.emit(
              'error',
              new Error('Encountered unskippable frame')
            )
        }
      } catch (err) {
        return this.emit('error', err)
      }

      _data = this.toNextFrame(_data)
    }

    if (_data.length) {
      this.frameBuffer = _data
    }

    if (chunks.length) {
      return this.processChunks(chunks, done)
    } else {
      return done()
    }
  }

  _flush(done: (err?: Error) => void) {
    const err = this.frameBuffer?.length
      ? new Error('Failed to decompress Snappy stream')
      : undefined
    done(err)
  }
}
